### PR TITLE
bundle.d/smartagent/collectd/mysql

### PR DESCRIPTION
**Pull request by rmfitzpatrick**
These changes introduce* initial smartagent/collectd/mysql discovery rules and an integration test. They also include windows `bundledFS` content differentiation since this is a native collectd monitor. I plan on introducing darwin build distinction in a subsequent PR since we don't have the agent bundle for that platform.

### CODE REVIEW

File: tests/testutils/kubeutils/kind_cluster.go [hunk 1]
```diff
@@ -171,9 +171,18 @@ func (k KindCluster) runKubectl(stdin io.Reader, args ...string) (stdOut, stdErr
 		// panic here to prevent swallowing what would have been a fatal error
 		panic(err)
 	})
+
+	defer func() {
+		// recovered from fatal kubectl command
+		if e := recover(); e != nil {
+			err = multierr.Combine(err, e.(error))
```
Comment (atoulme) [inline tests/testutils/kubeutils/kind_cluster.go:178]: Is that cast safe?

Comment (rmfitzpatrick) [reply tests/testutils/kubeutils/kind_cluster.go]: Not if the panic originated elsewhere w/o using an error value. Updated.

File: tests/testutils/kubeutils/kind_cluster.go [hunk 2]
```diff
@@ -153,28 +153,40 @@ func (k KindCluster) Apply(manifests string) (stdOut, stdErr bytes.Buffer, err e
 }
 
 func (k KindCluster) runKubectl(stdin io.Reader, args ...string) (stdOut, stdErr bytes.Buffer, err error) {
```
Comment (atoulme) [inline tests/testutils/kubeutils/kind_cluster.go:155]: It sounds like those changes are independent and can be peeled off as a separate PR to come in quickly. Not a big deal overall.

File: internal/confmapprovider/discovery/bundle/bundle_other.go [hunk 3]
```diff
@@ -12,6 +12,8 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+//go:build !windows
```
Comment (atoulme) [inline internal/confmapprovider/discovery/bundle/bundle_other.go:15]: Can you explain the split on Windows? Are some bundle rules not compatible with Windows and therefore you're excluding them from the build for Windows?

Comment (rmfitzpatrick) [reply internal/confmapprovider/discovery/bundle/bundle_other.go:15]: native collectd monitors like mysql aren't available on windows: https://github.com/signalfx/splunk-otel-collector/blob/main/pkg/receiver/smartagentreceiver/config.go#L39. Moving forward we'll need similar platform-specific bundles for other ones (like darwin).

Comment (rmfitzpatrick) [reply internal/confmapprovider/discovery/bundle/bundle_other.go:15]: the specifics of ^ are undefined for best experience and are currently based on what we provide agent bundles for.

File: tests/testutils/kubeutils/kind_cluster.go (hunk 2 above)
Comment (rmfitzpatrick) [reply tests/testutils/kubeutils/kind_cluster.go:155]: yeah they center around a bug I found writing these tests so threw them in the debugging process. I can break them out if you'd prefer but there's only one other consuming test.

**Review by crobert-1 — APPROVED** [review]

File: tests/receivers/smartagent/collectd-mysql/bundled_test.go [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
+
+	namespace, serviceAccount := cluster.createNamespaceAndServiceAccount()
+	mysqlUID := cluster.createMySQL("target.mysql", namespace, serviceAccount)
+	cluster.createClusterRoleAndRoleBinding(namespace, serviceAccount)
+	configMap := cluster.createConfigMap(namespace)
+	daemonSet := cluster.daemonSetManifest(namespace, serviceAccount, configMap)
+
+	var collectorName string
+	// wait for collector to run
+	require.Eventually(t, func() bool {
```
[the remark this answers is not in the document]
Comment (rmfitzpatrick) [reply tests/receivers/smartagent/collectd-mysql/bundled_test.go:58]: absolutely, I think the next case for k8s ideally to warrants a general collector helper like the ones for the binary and docker image that takes care of cluster creation, service account, daemonset, etc.

File: tests/testutils/kubeutils/kind_cluster.go (hunk 2 above)
Comment (atoulme) [reply tests/testutils/kubeutils/kind_cluster.go:155]: no, it's fine to have them here since they help with the PR, just conveying that in general having smaller PRs will help

File: internal/confmapprovider/discovery/bundle/bundle.d/receivers/smartagent-collectd-mysql.discovery.yaml [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+        - name: splunk.discovery
+      isolatedCollectd: true
+  status:
+    metrics:
+      successful:
+        - strict: mysql_octets.rx
+          first_only: true
+          log_record:
+            severity_text: info
+            body: collectd/mysql receiver is working!
```
Comment (theletterf) [inline internal/confmapprovider/discovery/bundle/bundle.d/receivers/smartagent-collectd-mysql.discovery.yaml:24]:
```suggestion
            body: MySQL receiver is working!
```

File: internal/confmapprovider/discovery/bundle/bundle.d/receivers/smartagent-collectd-mysql.discovery.yaml.tmpl [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+        - name: {{ defaultValue }}
+      isolatedCollectd: true
+  status:
+    metrics:
+      successful:
+        - strict: mysql_octets.rx
+          first_only: true
+          log_record:
+            severity_text: info
+            body: collectd/mysql receiver is working!
```
Comment (theletterf) [inline internal/confmapprovider/discovery/bundle/bundle.d/receivers/smartagent-collectd-mysql.discovery.yaml.tmpl:21]:
```suggestion
            body: MySQL receiver is working!
```

File: internal/confmapprovider/discovery/bundle/bundle.d/receivers/smartagent-collectd-mysql.discovery.yaml.tmpl [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+          log_record:
+            severity_text: info
+            body: The container is refusing MySQL connections.
+      partial:
+        - regexp: 'mysql plugin: Failed to connect to database .* at server .* Access denied for user .* \\(using password: YES\\)'
+          first_only: true
+          log_record:
+            severity_text: info
+            body: >-
+              Please ensure your user credentials are correctly specified with
```
Comment (theletterf) [inline internal/confmapprovider/discovery/bundle/bundle.d/receivers/smartagent-collectd-mysql.discovery.yaml.tmpl:35]:
```suggestion
              Make sure your user credentials are correctly specified using the
```

File: internal/confmapprovider/discovery/bundle/bundle.d/receivers/smartagent-collectd-mysql.discovery.yaml.tmpl [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+              `--set {{ configProperty "username" "<username>" }}` and
+              `--set {{ configProperty "password" "<password>" }}` or
```
Comment (theletterf) [inline internal/confmapprovider/discovery/bundle/bundle.d/receivers/smartagent-collectd-mysql.discovery.yaml.tmpl:37]:
```suggestion
              `--set {{ configProperty "password" "<password>" }}` command or the
```

File: internal/confmapprovider/discovery/bundle/bundle.d/receivers/smartagent-collectd-mysql.discovery.yaml.tmpl [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+              `{{ configPropertyEnvVar "username" "<username>" }}` and
+              `{{ configPropertyEnvVar "password" "<password>" }}` environment variables.
+        - regexp: 'mysql plugin: Failed to connect to database .* at server .* Unknown database'
+          first_only: true
+          log_record:
+            severity_text: info
+            body: >-
+              Please ensure your desired databases are correctly specified using the
+              `--set {{ configProperty "databases" "[{name: '<db-name-0>'}, {name: '<db-name-1>', username: '<username>', password: '<password>'}]" }}` command or the
+              `{{ configPropertyEnvVar "databases" "[{name: '<db-name-0>'}, {name: '<db-name-1>', username: '<username>', password: '<password>'}]" }}` environment variable.
```
Comment (theletterf) [inline internal/confmapprovider/discovery/bundle/bundle.d/receivers/smartagent-collectd-mysql.discovery.yaml.tmpl:47]:
```suggestion
              Make sure your MySQL databases are correctly specified using the
              `--set {{ configProperty "databases" "[{name: '<db-name-0>'}, {name: '<db-name-1>', username: '<username>', password: '<password>'}]" }}` command or the
              `{{ configPropertyEnvVar "databases" "[{name: '<db-name-0>'}, {name: '<db-name-1>', username: '<username>', password: '<password>'}]" }}` environment variable.
```

File: internal/confmapprovider/discovery/bundle/bundle.d/receivers/smartagent-collectd-mysql.discovery.yaml.tmpl [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+        - regexp: 'mysql plugin: Failed to connect to database .* at server .* Access denied for user .* to database'
+          first_only: true
+          log_record:
+            severity_text: info
+            body: >-
+              Please ensure your desired databases and auth information are correctly specified using the
+              `--set {{ configProperty "databases" "[{name: '<db-name-0>'}, {name: '<db-name-1>', username: '<username>', password: '<password>'}]" }}` command or the
+              `{{ configPropertyEnvVar "databases" "[{name: '<db-name-0>'}, {name: '<db-name-1>', username: '<username>', password: '<password>'}]" }}` environment variable.
```
Comment (theletterf) [inline internal/confmapprovider/discovery/bundle/bundle.d/receivers/smartagent-collectd-mysql.discovery.yaml.tmpl:55]:
```suggestion
              Make sure your MySQL databases and auth information are correctly specified using the
              `--set {{ configProperty "databases" "[{name: '<db-name-0>'}, {name: '<db-name-1>', username: '<username>', password: '<password>'}]" }}` command or the
              `{{ configPropertyEnvVar "databases" "[{name: '<db-name-0>'}, {name: '<db-name-1>', username: '<username>', password: '<password>'}]" }}` environment variable.
```

File: internal/confmapprovider/discovery/bundle/bundle.d/receivers/smartagent-collectd-mysql.discovery.yaml.tmpl (hunk 6 above)
Comment (rmfitzpatrick) [reply internal/confmapprovider/discovery/bundle/bundle.d/receivers/smartagent-collectd-mysql.discovery.yaml.tmpl]: I went ahead and adopted this one when fixing up the rules.
